### DISCUSSION TITLE
How do I use the results of supervised modeling on unseen data?

### DISCUSSION THREAD

**Discussion by JubinaMarie** [MaartenGr/BERTopic#1282] [Q&A]
First, thanks so, so much for BERTopic.  It's incredibly useful and even this newbie is mostly grasping it.  

I managed to run supervised topic-modeling on my training DF, but now I'm confused about how I would go about using the results to predict the topics in new data.  I understand in concept, but not how to make it work.  I've tried various things with no luck, including saving and loading the results of the supervised modeling to run on new data.  Thanks!

**Comment by JubinaMarie** [comment]
Hi again, I found this related discussion: https://github.com/MaartenGr/BERTopic/issues/278.  I am having the same issue mjavedgohar discusses.  I use fit_transform for training my supervised model.  The training data has 1,000 docs, 492 of which are "labeled XYZ topic" and 508 are "not XYZ topic."  Then I load the saved model and use it with "transform" on my test data, which is 800 documents.  The results when I do "get topic info" show again 492 xyz topic/508 not xyz topic, so clearly not working right.  It's like it's just ignoring the test data and spitting out results from the original training data.  What am I doing wrong?

**Reply by MaartenGr** [reply]
You are not doing anything wrong. The idea behind `.transform`, as is the same with scikit-learn, is that it does not modify the model at all. It is supposed to give back the predicted topics on your test data without actually modifying the model. The `.get_topic_info` method is used to show which topics were generated based on the training data, not the test data.

**Reply by MaartenGr** [reply]
If you want the model to update, you can look towards [online/incremental BERTopic](https://maartengr.github.io/BERTopic/getting_started/online/online.html) instead. Do note that it is quite a different procedure from supervised topic modeling though.

**Reply by JubinaMarie** [reply]
Hmm, I may not be doing anything wrong, but it's the wrong thing for what I'm trying to accomplish, which is to classify new data.  How do I apply the "trained" model to my new data?  You state in the supervised instructions that "a welcome side-effect is that we now have a classification algorithm that allows us to predict the topics of unseen data," but I don't know how to make that work.  Thank you so much for getting back!

**Reply by MaartenGr** [reply]
If I am not mistaken, what you are describing is that you want to train a model on data and discover a certain set of topics and then use that model on a new set of data to discover topics that were not found in the training data, correct? If so, then supervised topic modeling is not what you are looking for. 

> "a welcome side-effect is that we now have a classification algorithm that allows us to predict the topics of unseen data,"

It indeed allows us to predict the topics of unseen data but it will assign them only to topics that you have seen before. As mentioned, if you are looking for a model to discover different topics than what it was trained on, then  [online/incremental BERTopic](https://maartengr.github.io/BERTopic/getting_started/online/online.html) would be better suited since the topics change.

**Reply by JubinaMarie** [reply]
"It indeed allows us to predict the topics of unseen data but it will assign them only to topics that you have seen before."  Actually, this is exactly what I am trying to do.  I already identified my topics in the training set manually for supervised modeling.  Now I want to assign only those same topics to new data.  Just can't figure out how to do the assigning of the topics on new data.

**Reply by MaartenGr** [reply]
> Now I want to assign only those same topics to new data. 

In that case, you would only need to run `.transform` to assign the same topics to new data. The output of `.transform` are exactly those topics.

**Reply by JubinaMarie** [reply]
Thank you for pointing me in the right direction.  I was able to make it work.
